src/MOLFiniteDifference/MOL_discretization.jl

```diff
@@ -126,18 +135,47 @@ function SciMLBase.symbolic_discretize(pdesys::ModelingToolkit.PDESystem,discret
-        #reverse_weights(i,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, [space[j][i[j]-1],space[j][i[j]]])
-        #upwinding_rules = [@rule(*(~~a,(Differential(nottime[j]))(u),~~b) => IfElse.ifelse(*(~~a..., ~~b...,)>0,
-        #                        *(~~a..., ~~b..., dot(reverse_weights(i,j),depvars[k][central_neighbor_idxs(i,j)[1:2]])),
-        #                        *(~~a..., ~~b..., dot(forward_weights(i,j),depvars[k][central_neighbor_idxs(i,j)[2:3]]))))
-        #                        for j in 1:nspace, k in 1:length(pdesys.depvars)]
-    
-        rules = vcat(vec(central_deriv_rules),valrules)
+        forward_weights(i,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, [space[j][i[j]],space[j][i[j]+1]])
+        reverse_weights(i,j) = DiffEqOperators.calculate_weights(discretization.upwind_order, 0.0, [space[j][i[j]-1],space[j][i[j]]])
+        # upwinding_rules = [@rule(*(~~a,$(Differential(nottime[j]))(u),~~b) => IfElse.ifelse(*(~~a..., ~~b...,)>0,
+        #                         *(~~a..., ~~b..., dot(reverse_weights(i,j),depvars[k][central_neighbor_idxs(i,j)[1:2]])),
+        #                         *(~~a..., ~~b..., dot(forward_weights(i,j),depvars[k][central_neighbor_idxs(i,j)[2:3]]))))
+        #                         for j in 1:nspace, k in 1:length(pdesys.depvars)]
+
+        ## Discretization of non-linear laplacian. 
+        # d/dx( a du/dx ) ~ (a(x+1/2) * (u[i+1] - u[i]) - a(x-1/2) * (u[i] - u[i-1]) / dx^2
+        b1(i, j, k) = dot(reverse_weights(i, j), depvars[k][central_neighbor_idxs(i, j)[1:2]]) / dxs[j]
+        b2(i, j, k) = dot(forward_weights(i, j), depvars[k][central_neighbor_idxs(i, j)[2:3]]) / dxs[j]
+        # TODO: improve interpolation of g(x) = u(x) for calculating u(x+-dx/2)
+        g(i, j, k, l) = sum([depvars[k][central_neighbor_idxs(i, j)][s] for s in (l == 1 ? [2,3] : [1,2])]) / 2.
+        # iv_mid returns middle space values. E.g. x(i-1/2) or y(i+1/2).
+        iv_mid(i, j, l) = (space[j][i[j]] + space[j][i[j]+l]) / 2.0 
+        # Dependent variable rules
+        r_mid_dep(i, j, k, l) = [pdesys.depvars[k] => g(i, j, k, l) for k in 1:length(pdesys.depvars)]
+        # Independent variable rules
+        r_mid_indep(i, j, l) = [nottime[j] => iv_mid(i, j, l) for j in 1:length(nottime)]
+        # Replacement rules: new approach
+        rules = [@rule ($(Differential(iv))(*(~~a, $(Differential(iv))(dv), ~~b))) =>
+                 dot([Num(substitute(substitute(*(~~a..., ~~b...), r_mid_dep(II, j, k, -1)), r_mid_indep(II, j, -1))),
+                      Num(substitute(substitute(*(~~a..., ~~b...), r_mid_dep(II, j, k, 1)), r_mid_indep(II, j, 1)))],
+                     [-b1(II, j, k), b2(II, j, k)])
+                 for (j, iv) in enumerate(nottime) for (k, dv) in enumerate(pdesys.depvars)]
+        rhs_arg = (SymbolicUtils.operation(eq.rhs) == +) ? SymbolicUtils.arguments(eq.rhs) : [eq.rhs]
```

Comment:
why do you need to explicitly specialize on `+`?

Reply:
My goal is to run classical convection-diffusion equations. To do so, in this PR I addressed summations involving nonlinear Laplacians, needed in diffusive terms.